### PR TITLE
Versioned table entries (reorg part 2/4)

### PR DESCRIPTION
**Pull request by AshtonStephens**
## Description

Closes: #N/A

Part of the reorg handling PR, but broken up.

## Changes

Adds versioned table entries with some generic handling.

## Testing Information

Tested with integration tests.

## Checklist:

- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: emily/handler/src/common/error.rs [hunk 1]
```diff
@@ -172,7 +194,13 @@ impl From<SdkError<BatchWriteItemError>> for Error {
 }
 impl From<SdkError<UpdateItemError>> for Error {
     fn from(err: SdkError<UpdateItemError>) -> Self {
-        Error::Debug(format!("SdkError<UpdateItemError> - {err:?}"))
+        match err.into_service_error() {
+            // Note, this assumes that any conditional check that fails fails because
+            // there's a version conflict. This isn't necessarily true but is a good
+            // simplifying assumption.
```
Comment (djordon) [inline emily/handler/src/common/error.rs:200]: Do we know of any cases where this isn't the case?

File: emily/handler/src/database/entries/mod.rs [hunk 2]
```diff
@@ -413,6 +413,70 @@ pub(crate) trait TableIndexTrait {
...
+    /// Increment version.
+    fn increment_version(&mut self);
+}
+
+/// Index trait for a versioned thing.
+pub(crate) trait VersionedTableIndexTrait: TableIndexTrait
+where
+    Self::Entry: VersionedEntryTrait,
+{
+    /// Put generic table entry but add version a check.
```
Comment (djordon) [inline emily/handler/src/database/entries/mod.rs:431]: nit: typo

File: emily/handler/src/database/entries/mod.rs [hunk 3]
```diff
@@ -413,6 +413,70 @@ pub(crate) trait TableIndexTrait {
...
+    ///
+    /// Note that this assumes that the entry has a version field called "Version"
+    /// and that the entry being emplaced has the "Version" field CURRENTLY set to
+    /// the EXPECTED version of the entry to update. This function will automatically
+    /// increment the version of the entry on PUT.
```
Comment (djordon) [inline emily/handler/src/database/entries/mod.rs:436]: Is this still true?

**Review by djordon — APPROVED** [review]
Had a few minor questions, but it looks good.

File: emily/handler/src/database/entries/mod.rs (hunk 3 above)
Comment (AshtonStephens) [reply emily/handler/src/database/entries/mod.rs:436]: No, nice catch!

File: emily/handler/src/database/entries/mod.rs (hunk 2 above)
Comment (AshtonStephens) [reply emily/handler/src/database/entries/mod.rs:431]: Fixed.

File: emily/handler/src/common/error.rs (hunk 1 above)
Comment (AshtonStephens) [reply emily/handler/src/common/error.rs:200]: If you make any kind of conditional check this would fail and it would say it's a version conflict.

**Pull request merged**